vortex-ipc/src/reader.rs

```diff
@@ -135,84 +137,121 @@ impl<'a, R: Read> StreamArrayReader<'a, R> {
+                vortex_bail!("Indices must be positive")
+            }
         }
 
         if self.row_offset != 0 {
             vortex_bail!("Stream has already been (at least partially) consumed")
         }
 
-        let mut chunks = Vec::new();
+        Ok(TakeIterator {
+            reader: self,
+            indices,
+            row_offset: 0,
+        })
+    }
+}
 
-        let mut row_offset = self.row_offset;
+pub struct TakeIterator<'a, R: Read> {
+    reader: StreamArrayReader<'a, R>,
+    indices: &'a Array<'a>,
+    row_offset: usize,
+}
 
-        // Continue reading batches from the stream until we run out
-        while let Some(batch) = self.next()? {
-            let left =
-                search_sorted::<usize>(indices, row_offset, SearchSortedSide::Left)?.to_index();
-            let right =
-                search_sorted::<usize>(indices, row_offset + batch.len(), SearchSortedSide::Left)?
-                    .to_index();
-            // NB: if we're done, make sure to consume the rest of the
-            //  array to leave the iterator in a good state
-            if left == indices.len() {
-                continue;
+/// NB: if the StreamArrayReader expires without being fully-consumed by the underlying iterator,
+/// it will leave the underlying messages stream in an unreadable state and free up a mutable ref
+/// to that stream, allowing someone else to try to issue a read that will inevitably fail.
+/// For this reason, we force full consumption of the reader when it goes out of scope.
+impl<'a, R: Read> Drop for StreamArrayReader<'a, R> {
+    fn drop(&mut self) {
+        // NB: can't catch_unwind/unwrap here because &mut self can't be made UnwindSafe
+        loop {
+            let next = self.next();
+            match next {
+                Ok(result) => {
+                    if result.is_none() {
+                        break;
+                    }
+                }
+                Err(err) => {
+                    error!("Error consuming StreamArrayReader in destructor: {:?}", err);
+                    break;
+                }
             }
+        }
+    }
+}
+
+/// NB: if the TakeIterator expires without being fully-consumed by the underlying iterator,
+/// it will trigger the Drop impl on the StreamArrayReader, which will consume the rest of the
+/// messages for that array. This is necessary to ensure that the underlying stream is in a
+/// consistent state when the StreamArrayReader is dropped, but it also is not free. If users wish
+/// to avoid this work happening at exipry, users can just consume the rest of the iterator
+/// themselves when they see fit.
+impl<'a, R: Read> FallibleIterator for TakeIterator<'a, R> {
+    type Item = OwnedArray;
+    type Error = VortexError;
+
+    fn next(&mut self) -> VortexResult<Option<Self::Item>> {
+        if self.indices.is_empty() {
+            return Ok(None);
+        }
+        while let Some(batch) = self.reader.next().unwrap() {
```

Comment:
`.unwrap()` can be `?` here, I think, which will propagate the error instead of panicking